lib/netext/dialer.go

```diff
@@ -149,6 +142,68 @@ func (d *Dialer) GetTrail(
 	}
 }
 
+func (d *Dialer) getDialAddr(addr string) (string, error) {
+	remote, err := d.findRemote(addr)
+	if err != nil {
+		return "", err
+	}
+
+	for _, ipnet := range d.Blacklist {
+		if ipnet.Contains(remote.IP) {
+			return "", BlackListedIPError{ip: remote.IP, net: ipnet}
+		}
+	}
+
+	return remote.String(), nil
+}
+
+func (d *Dialer) findRemote(addr string) (*lib.HostAddress, error) {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	remote, err := d.getCachedHost(addr, host, port)
+	if err != nil || remote != nil {
+		return remote, err
+	}
+
+	ip := net.ParseIP(host)
+
+	if ip == nil {
+		ip, err = d.Resolver.FetchOne(host)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if ip == nil {
+		return nil, errors.Errorf("lookup %s: no such host", host)
+	}
+
+	return lib.NewHostAddress(ip, port)
+}
+
+func (d *Dialer) getCachedHost(addr, host, port string) (*lib.HostAddress, error) {
+	// lookup for full address defined in Hosts option before trying to resolve DNS.
+	if remote, ok := d.Hosts[addr]; ok {
+		return remote, nil
+	}
+
+	// lookup for host defined in Hosts option before trying to resolve DNS.
+	if remote, ok := d.Hosts[host]; ok {
+		if remote.Port == 0 && port != "" {
+			newPort, err := strconv.Atoi(port)
+			if err != nil {
+				return nil, err
+			}
+			remote.Port = newPort
+		}
+		return remote, nil
+	}
```

Comment:
As the test has shown here you get a pointer to the remote which you then ... change, which definitely should never happen given that this is the configuration ;)